src/ModelContextProtocol/Auth/McpClientExtensions.cs

```diff
@@ -0,0 +1,122 @@
+            ClientId = clientId,
+            ClientName = clientName,
+            Scopes = scopes?.ToList(),
+            AuthorizationHandler = handler
+        };
+        
+        _clientConfigs[httpClient] = config;
+    }
+    
+    /// <summary>
+    /// Gets the authorization configuration for the HTTP client.
+    /// </summary>
+    /// <param name="httpClient">The HTTP client.</param>
+    /// <returns>The authorization configuration, or null if not configured.</returns>
+    public static AuthorizationConfig? GetAuthorizationConfig(this HttpClient httpClient)
+    {
+        _clientConfigs.TryGetValue(httpClient, out var config);
+        return config;
+    }
+    
+    /// <summary>
+    /// Handles a 401 Unauthorized response from an MCP server.
+    /// </summary>
+    /// <param name="httpClient">The HTTP client.</param>
+    /// <param name="response">The HTTP response with the 401 status code.</param>
+    /// <returns>The OAuth token response if authentication was successful.</returns>
+    public static async Task<OAuthToken> HandleUnauthorizedResponseAsync(
+        this HttpClient httpClient,
+        HttpResponseMessage response)
+    {
+        if (response.StatusCode != System.Net.HttpStatusCode.Unauthorized)
+        {
+            throw new ArgumentException("The response status code must be 401 Unauthorized.", nameof(response));
+        }
+        
+        // Get the WWW-Authenticate header
+        var wwwAuthenticateHeader = response.Headers.WwwAuthenticate.ToString();
+        if (string.IsNullOrEmpty(wwwAuthenticateHeader))
+        {
+            throw new InvalidOperationException("The response does not contain a WWW-Authenticate header.");
+        }
+        
+        // Get the authorization configuration
+        var config = httpClient.GetAuthorizationConfig();
+        if (config == null)
+        {
+            throw new InvalidOperationException("The HTTP client has not been configured for authorization handling. Call ConfigureAuthorizationHandler() first.");
+        }
+        
+        // Create OAuthAuthenticationService - use appropriate constructor based on whether we have a handler
+        OAuthAuthenticationService authService = config.AuthorizationHandler != null
+            ? new OAuthAuthenticationService(config.AuthorizationHandler)
+            : new OAuthAuthenticationService();
+        
+        // Get resource URI
+        var resourceUri = response.RequestMessage?.RequestUri ?? throw new InvalidOperationException("Request URI is not available.");
+        
+        // Start the authentication flow
+        var tokenResponse = await authService.HandleAuthenticationAsync(
+            resourceUri,
+            wwwAuthenticateHeader,
+            config.RedirectUri,
+            config.ClientId,
+            config.ClientName,
+            config.Scopes);
+        
+        // Attach the access token to future requests
+        httpClient.AttachToken(tokenResponse.AccessToken);
```

Comment:
Looks like this may not handle token expiry and refresh, as in the token could expire and get stuck on the http client instance, never succeeding after expiration.